File: bikes/Code/Model/ride.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using PetaPoco;

namespace Bikes.Model
{
    [PetaPoco.TableName("ride")]
    [PetaPoco.PrimaryKey("id")]
    public class Ride
    {
        public int id { get; set; }

        public int bike_id { get; set; }
        public int rider_id { get; set; }
        public int route_id { get; set; }

        [PetaPoco.ResultColumn]
        public String bike { get; set; }
        [PetaPoco.ResultColumn]
        public String rider { get; set; }
        [PetaPoco.ResultColumn]
        public String route { get; set; }
        [PetaPoco.ResultColumn]
        public int rate { get; set; }

        [PetaPoco.Ignore]
        public double amount
        {
            get
            {
                return (distance * rate) / 100f;
            }
        }


        public double distance { get; set; }
        public String notes { get; set; }
        public DateTime ride_date { get; set; }

        public static List<Ride> getRides()
        {
            Database db = new PetaPoco.Database("bikes-clunie");
            List<Ride> rides = db.Fetch<Ride>(
                PetaPoco.Sql.Builder
                    .Append("SELECT a.id, a.route_id, a.rider_id, a.bike_id, a.distance, a.ride_date, a.notes, a.last_updated,")
                    .Append("b.name AS route, c.name AS rider, c.rate AS rate, d.name AS bike")
                    .Append("FROM ride a, route b, rider c, bike d")
                    .Append("WHERE a.route_id = b.id AND a.rider_id = c.id AND a.bike_id = d.id"));

            return rides;
        }

        public static Ride getRide(int id)
        {
            Database db = new PetaPoco.Database("bikes-clunie");
            Ride ride = db.FirstOrDefault<Ride>(
                PetaPoco.Sql.Builder
                    .Append("SELECT a.id, a.route_id, a.rider_id, a.bike_id, a.distance, a.ride_date, a.notes, a.last_updated,")
                    .Append("b.name AS route, c.name AS rider, c.rate AS rate, d.name AS bike")
                    .Append("FROM ride a, route b, rider c, bike d")
                    .Append("AND a.id = @0", id));

            return ride;
        }

        public static void deleteRide(int id)
        {
            Database db = new PetaPoco.Database("bikes-clunie");
            db.Execute("DELETE FROM ride WHERE id = @0", id);
        }

        public void save()
        {
            Database db = new PetaPoco.Database("bikes-clunie");
            db.Save(this);
        }
    }
}

File: bikes/Source/app/ride.js
angular.module('bikesApp')

.controller('rideListController', ["$scope", "$state", "model", function (scope, state, model) {

    scope.model = model;

    scope.add = function () {
        state.go("rideAdd");
    };

}])

.controller('rideAddController', ["$scope", "$state", "$stateParams", "model", "Ride", "Rider", "Route", "Bike",
    function (scope, state, stateParams, model, Ride, Rider, Route, Bike) {

        scope.dateOptions = {
            showWeeks: false
        };

        scope.openDatePicker = function () {
            scope.datePickerOpen = true;
        };

        scope.routeCaption = function (route) {
            if (route.id === 1) {
                return route.name;
            } else {
                return route.name + ' (' + route.distance + ' miles)';
            }
        };

        var goBack = function () {
            if (stateParams.riderId) {
                state.go("userHome");
            } else {
                state.go("rideList");
            }
        };

        var getRider = function (id) {
            var i;
            for (i = 0; i < scope.riders.length; i++) {
                if (scope.riders[i].id === id) {
                    return scope.riders[i];
                }
            }
        }

        var getRoute = function (id) {
            var i;
            for (i = 0; i < scope.routes.length; i++) {
                if (scope.routes[i].id === id) {
                    return scope.routes[i];
                }
            }
        }

        scope.routes = Route.query();
        scope.bikes  = Bike.query();
        scope.riders = Rider.query();

        scope.ride = new Ride();
        scope.ride.distance = 0;
        scope.ride.ride_date = new Date();

        if (model.data.currentRider) {
            scope.ride.payable = model.data.currentRider.default_payable;
        }

        scope.ride.bonus = 0;
        scope.ride.rider_id = stateParams.riderId ? parseInt(stateParams.riderId, 10) : 0;
        scope.ride.route_id = 1;
        scope.ride.bike_id = 1;

        scope.riderChanged = function () {
            scope.rider = getRider(scope.ride.rider_id);
            scope.ride.payable = scope.rider.default_payable;
        };

        scope.calculateReward = function () {
            var distance;
            var reward;

            if (!scope.rider) {
                return 0;
            }

            if (!scope.ride.payable) {
                reward = 0;
            } else {
                //see if there is a route selected
                if (scope.ride.route_id > 1) {
                    distance = getRoute(scope.ride.route_id).distance;

                    if (scope.ride.returnRide) {
                        distance = distance * 2;
                    }
                } else {
                    distance = scope.ride.distance;
                }

                reward = scope.ride.bonus + (distance * scope.rider.rate) / 100;
            }
            return reward.toFixed(2);
        };

        scope.saveForm = function () {

            //save the ride
            scope.ride.$save(
                //on success
                function () {
                    model.refresh();
                    goBack();
                },
                //on failure
                function () {
                    alert("Failed to save new ride.");
                    goBack();
                }
            );
        };

        scope.cancel = function () {
            goBack();
        };
    }
])

.controller('rideEditController', ["$scope", "$state", "$stateParams", "model", "Ride", function (scope, state, stateParams, model, Ride) {

    scope.ride = Ride.get({ id: stateParams.id });

    var goBack = function () {
        if (stateParams.riderId) {
            state.go("userHome");
        } else {
            state.go("rideList");
        }
    };

    scope.saveForm = function () {
        scope.ride.$save(function () {
            model.refresh();
            goBack();
        });
    };

    scope.deleteRide = function () {
       Ride.trash(
            { id: stateParams.id },
            function (data) {
                if (data.status === "OK") {
                    model.refresh();
                    goBack();
                } else if (data.status === "ALREADY_PAID") {
                    scope.message = "Rides that have been paid cannot be deleted.";
                } else {
                    scope.message = "There was a problem deleting this ride.";
                }
            },
            function (response) {
                alert("failed to delete ride - " + response.status);
            }
        );
    };

    scope.cancel = function () {
        goBack();
    };
}]);
